File: franka_spine/franka_spine_examples/scripts/spine_client_example.py
#!/usr/bin/env python3
# Copyright (c) 2026 Franka Robotics GmbH
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""Example client that exercises the Franka Spine action/service server."""

from franka_spine_msgs.action import MoveAbsolute
from franka_spine_msgs.srv import (
    GetParameters,
    GetPosition,
    GetSpineState,
    SwitchOff,
    SwitchOn,
)
import rclpy
from rclpy.action import ActionClient
from rclpy.node import Node


class SpineClient(Node):
    """Minimal client for the Franka Spine server."""

    def __init__(self):
        super().__init__('spine_client')
        self.state_client = self.create_client(GetSpineState, '/franka_spine_node/get_state')
        self.position_client = self.create_client(GetPosition, '/franka_spine_node/get_position')
        self.switch_on_client = self.create_client(SwitchOn, '/franka_spine_node/switch_on')
        self.switch_off_client = self.create_client(SwitchOff, '/franka_spine_node/switch_off')
        self.get_parameters_client = self.create_client(
            GetParameters, '/franka_spine_node/get_parameters_spine'
        )
        self.move_client = ActionClient(self, MoveAbsolute, '/franka_spine_node/move_absolute')
        self.get_logger().info('Waiting for spine services and action server...')
        try:
            if not self.state_client.wait_for_service(timeout_sec=5.0):
                self.get_logger().warning('GetState service not available after timeout')
            if not self.position_client.wait_for_service(timeout_sec=5.0):
                self.get_logger().warning('GetPosition service not available after timeout')
            if not self.switch_on_client.wait_for_service(timeout_sec=5.0):
                self.get_logger().warning('SwitchOn service not available after timeout')
            if not self.switch_off_client.wait_for_service(timeout_sec=5.0):
                self.get_logger().warning('SwitchOff service not available after timeout')
            if not self.get_parameters_client.wait_for_service(timeout_sec=5.0):
                self.get_logger().warning('GetParameters service not available after timeout')
            if not self.move_client.wait_for_server(timeout_sec=5.0):
                self.get_logger().warning('Move action server not available after timeout')
        except Exception as e:
            self.get_logger().warning(f'Exception while waiting for services: {e}')

    def get_state(self):
        """Query the current spine state."""
        future = self.state_client.call_async(GetSpineState.Request())
        rclpy.spin_until_future_complete(self, future)
        result = future.result()
        self.get_logger().info(f'State: {result.state}, success: {result.success}')
        return result

    def get_position(self):
        """Query the current position in metres."""
        future = self.position_client.call_async(GetPosition.Request())
        rclpy.spin_until_future_complete(self, future)
        result = future.result()
        self.get_logger().info(f'Position: {result.position:.4f} m, success: {result.success}')
        return result

    def switch_on(self):
        """Switch the spine on."""
        future = self.switch_on_client.call_async(SwitchOn.Request())
        rclpy.spin_until_future_complete(self, future)
        result = future.result()
        self.get_logger().info(f'Switch on: {result.message}')
        return result

    def switch_off(self):
        """Switch the spine off."""
        future = self.switch_off_client.call_async(SwitchOff.Request())
        rclpy.spin_until_future_complete(self, future)
        result = future.result()
        self.get_logger().info(f'Switch off: {result.message}')
        return result

    def get_parameters(self):
        """Retrieve spine parameters (user limits)."""
        future = self.get_parameters_client.call_async(GetParameters.Request())
        rclpy.spin_until_future_complete(self, future)
        result = future.result()
        if result.success:
            limits = result.parameters.user_limits
            self.get_logger().info(
                f'User limits: [{limits.lower_limit:.4f}, {limits.upper_limit:.4f}] m'
            )
        return result

    def move_to(
        self,
        position,
        velocity=0.1,
        acceleration=0.2,
        deceleration=0.2,
    ):
        """Send a MoveAbsolute goal and wait for the result."""
        self.move_client.wait_for_server()
        goal = MoveAbsolute.Goal()
        goal.position = position
        goal.velocity = velocity
        goal.acceleration = acceleration
        goal.deceleration = deceleration

        future = self.move_client.send_goal_async(goal, feedback_callback=self._feedback_callback)
        rclpy.spin_until_future_complete(self, future)

        goal_handle = future.result()
        if not goal_handle.accepted:
            self.get_logger().error('Goal rejected')
            return None

        result_future = goal_handle.get_result_async()
        rclpy.spin_until_future_complete(self, result_future)
        return result_future.result().result

    def _feedback_callback(self, feedback_msg):
        """Log position feedback during motion."""
        self.get_logger().info(f'Position: {feedback_msg.feedback.current_position:.4f} m')


def main():
    rclpy.init()
    client = SpineClient()

    print('--- Franka Spine Client Example ---')

    # Switch on
    client.switch_on()

    # Query state and position
    client.get_state()
    client.get_position()
    client.get_parameters()

    result = client.move_to(
        position=0.4,
        velocity=0.1,
        acceleration=0.2,
        deceleration=0.2,
    )
    if result and result.success:
        client.get_logger().info('Motion 1 completed successfully')

    result = client.move_to(
        position=0.05,
        velocity=0.1,
        acceleration=0.2,
        deceleration=0.2,
    )
    if result and result.success:
        client.get_logger().info('Motion 2 completed successfully')

    # Switch off
    client.switch_off()

    client.destroy_node()
    rclpy.shutdown()


if __name__ == '__main__':
    main()

File: franka_spine/franka_spine_examples/src/spine_client_example.cpp
// Copyright (c) 2026 Franka Robotics GmbH
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include "franka_spine_examples/spine_client_example.hpp"

#include <atomic>
#include <chrono>
#include <future>
#include <stdexcept>

using namespace std::chrono_literals;

static constexpr const char* kGetStateServiceName = "/franka_spine_node/get_state";
static constexpr const char* kGetPositionServiceName = "/franka_spine_node/get_position";
static constexpr const char* kSwitchOnServiceName = "/franka_spine_node/switch_on";
static constexpr const char* kSwitchOffServiceName = "/franka_spine_node/switch_off";
static constexpr const char* kMoveActionName = "/franka_spine_node/move_absolute";
static constexpr std::chrono::seconds kDefaultTimeout{5s};

namespace {

template <typename T>
auto waitForService(const rclcpp::Logger& logger,
                    const typename rclcpp::Client<T>::SharedPtr& client,
                    const std::string& name) -> bool {
  if (!client->wait_for_service(kDefaultTimeout)) {
    RCLCPP_ERROR(logger, "Service '%s' not available.", name.c_str());
    return false;
  }

  return true;
}

}  // namespace

namespace franka_spine_examples {

SpineClientExample::SpineClientExample(const rclcpp::NodeOptions& options)
    : rclcpp::Node("spine_client_cpp", options) {
  state_client_ = create_client<GetSpineState>(kGetStateServiceName);
  position_client_ = create_client<GetPosition>(kGetPositionServiceName);
  switch_on_client_ = create_client<SwitchOn>(kSwitchOnServiceName);
  switch_off_client_ = create_client<SwitchOff>(kSwitchOffServiceName);
  move_client_ = rclcpp_action::create_client<MoveAbsolute>(this, kMoveActionName);
  if (!waitForService<GetSpineState>(get_logger(), state_client_, kGetStateServiceName) ||
      !waitForService<GetPosition>(get_logger(), position_client_, kGetPositionServiceName) ||
      !waitForService<SwitchOn>(get_logger(), switch_on_client_, kSwitchOnServiceName) ||
      !waitForService<SwitchOff>(get_logger(), switch_off_client_, kSwitchOffServiceName)) {
    throw std::runtime_error("Timed out waiting for Franka Spine services.");
  }

  if (!move_client_->wait_for_action_server(kDefaultTimeout)) {
    throw std::runtime_error("Timed out waiting for Franka Spine MoveAbsolute action server.");
  }
}

auto SpineClientExample::run() -> void {
  getState();
  getPosition();
  switchOn();
  moveTo(0.4, 0.1, 0.2, 0.2);
  moveTo(0.05, 0.1, 0.2, 0.2);
  switchOff();

  RCLCPP_INFO(get_logger(), "Spine client example completed.");
}

auto SpineClientExample::getState() -> bool {
  auto future = state_client_->async_send_request(std::make_shared<GetSpineState::Request>());
  if (rclcpp::spin_until_future_complete(shared_from_this(), future) !=
      rclcpp::FutureReturnCode::SUCCESS) {
    RCLCPP_ERROR(get_logger(), "Failed to call get_state.");
    return false;
  }

  auto result = future.get();
  RCLCPP_INFO(get_logger(), "State: %s, success: %s", result->state.c_str(),
              result->success ? "true" : "false");
  return result->success;
}

auto SpineClientExample::getPosition() -> bool {
  auto future = position_client_->async_send_request(std::make_shared<GetPosition::Request>());
  if (rclcpp::spin_until_future_complete(shared_from_this(), future) !=
      rclcpp::FutureReturnCode::SUCCESS) {
    RCLCPP_ERROR(get_logger(), "Failed to call get_position.");
    return false;
  }

  auto result = future.get();
  RCLCPP_INFO(get_logger(), "Position: %.4f m, success: %s", result->position,
              result->success ? "true" : "false");
  return result->success;
}

auto SpineClientExample::switchOn() -> bool {
  auto future = switch_on_client_->async_send_request(std::make_shared<SwitchOn::Request>());
  if (rclcpp::spin_until_future_complete(shared_from_this(), future) !=
      rclcpp::FutureReturnCode::SUCCESS) {
    RCLCPP_ERROR(get_logger(), "Failed to call switch_on.");
    return false;
  }

  auto result = future.get();
  RCLCPP_INFO(get_logger(), "Switch on: %s", result->message.c_str());
  return result->success;
}

auto SpineClientExample::switchOff() -> bool {
  auto future = switch_off_client_->async_send_request(std::make_shared<SwitchOff::Request>());
  if (rclcpp::spin_until_future_complete(shared_from_this(), future) !=
      rclcpp::FutureReturnCode::SUCCESS) {
    RCLCPP_ERROR(get_logger(), "Failed to call switch_off.");
    return false;
  }

  auto result = future.get();
  RCLCPP_INFO(get_logger(), "Switch off: %s", result->message.c_str());
  return result->success;
}

auto SpineClientExample::moveTo(double position,
                                double velocity,
                                double acceleration,
                                double deceleration) -> bool {
  auto goal = MoveAbsolute::Goal();
  goal.position = position;
  goal.velocity = velocity;
  goal.acceleration = acceleration;
  goal.deceleration = deceleration;

  RCLCPP_INFO(get_logger(),
              "Sending move goal: position=%.4f m, velocity=%.4f m/s, "
              "acceleration=%.4f m/s², deceleration=%.4f m/s²",
              position, velocity, acceleration, deceleration);

  rclcpp_action::Client<MoveAbsolute>::SendGoalOptions options;
  options.goal_response_callback =
      [this](std::shared_ptr<rclcpp_action::ClientGoalHandle<MoveAbsolute>> handle) {
        handleMoveGoalResponse(handle);
      };
  options.feedback_callback = [this](
                                  std::shared_ptr<rclcpp_action::ClientGoalHandle<MoveAbsolute>>,
                                  const std::shared_ptr<const MoveAbsolute::Feedback>& feedback) {
    handleMoveFeedback(feedback);
  };

  std::atomic<bool> motion_finished{false};
  bool motion_success = false;
  options.result_callback =
      [&](const rclcpp_action::ClientGoalHandle<MoveAbsolute>::WrappedResult& result) {
        handleMoveResult(result);
        motion_success =
            (result.code == rclcpp_action::ResultCode::SUCCEEDED && result.result->success);
        motion_finished.store(true);
      };

  auto future_handle = move_client_->async_send_goal(goal, options);
  const auto& goal_handle = future_handle.get();
  if (!goal_handle) {
    RCLCPP_ERROR(get_logger(), "Move goal was rejected by server.");
    return false;
  }

  while (!motion_finished.load()) {
    rclcpp::sleep_for(10ms);
  }

  return motion_success;
}

auto SpineClientExample::handleMoveGoalResponse(
    const std::shared_ptr<rclcpp_action::ClientGoalHandle<MoveAbsolute>>& goal_handle) -> void {
  if (!goal_handle) {
    RCLCPP_ERROR(get_logger(), "Move goal was rejected.");
  } else {
    RCLCPP_INFO(get_logger(), "Move goal accepted.");
  }
}

auto SpineClientExample::handleMoveFeedback(
    const std::shared_ptr<const MoveAbsolute::Feedback>& feedback) -> void {
  RCLCPP_INFO(get_logger(), "Position: %.4f m", feedback->current_position);
}

auto SpineClientExample::handleMoveResult(
    const rclcpp_action::ClientGoalHandle<MoveAbsolute>::WrappedResult& result) -> void {
  switch (result.code) {
    case rclcpp_action::ResultCode::SUCCEEDED:
      if (result.result->success) {
        RCLCPP_INFO(get_logger(), "Motion completed successfully.");
      } else {
        RCLCPP_ERROR(get_logger(), "Motion finished with error: %s", result.result->error.c_str());
      }
      break;
    case rclcpp_action::ResultCode::ABORTED:
      RCLCPP_ERROR(get_logger(), "Motion aborted: %s", result.result->error.c_str());
      break;
    case rclcpp_action::ResultCode::CANCELED:
      RCLCPP_WARN(get_logger(), "Motion canceled.");
      break;
    default:
      RCLCPP_ERROR(get_logger(), "Unknown result code.");
      break;
  }
}

}  // namespace franka_spine_examples

int main(int argc, char** argv) {
  rclcpp::init(argc, argv);
  auto node = std::make_shared<franka_spine_examples::SpineClientExample>();
  auto run_node = std::async(std::launch::async, [&node]() { node->run(); });

  while (run_node.wait_for(1ms) != std::future_status::ready) {
    rclcpp::spin(node);
  }

  rclcpp::shutdown();
  return 0;
}
